src/ATen/native/xpu/sycl/LayerNormKernels.cpp

```diff
@@ -268,45 +169,297 @@ class LayerNormBackward : public NormBackward<scalar_t, mean_t, weight_t> {
   int64_t numel;
 };
 
-template <typename scalar_t, typename mean_t, typename weight_t>
+constexpr int vec_size =
+    4; // we could make it dependent on dtype, but that would lead to different
+       // results between float and low-p types
+
+// Checks alignment of buffers for using vectorized loads / stores
+template <typename T>
+bool can_vectorize(const T* ptr, int alignment) {
+  uint64_t addr = reinterpret_cast<uint64_t>(ptr);
+  return addr % alignment == 0;
+};
+
+struct WelfordDataLN {
+  float mean;
+  float sigma2;
+  float count;
+  WelfordDataLN() : mean(0.f), sigma2(0.f), count(0.f) {}
+  WelfordDataLN(float mean, float sigma2, float count)
+      : mean(mean), sigma2(sigma2), count(count) {}
+};
+
+template <typename U>
+WelfordDataLN WelfordOnlineSum(const U val, const WelfordDataLN& curr_sum) {
+  U delta = val - curr_sum.mean;
+  U new_count = curr_sum.count + 1.f;
+  U new_mean = curr_sum.mean +
+      delta * (1.f / new_count); // proper division is slow, this is less
+                                 // accurate but noticeably faster
+  return WelfordDataLN(
+      new_mean, curr_sum.sigma2 + delta * (val - new_mean), new_count);
+}
+
+WelfordDataLN WelfordCombine(
+    const WelfordDataLN dataB,
+    const WelfordDataLN dataA) {
+  using U = decltype(dataB.count);
+  U delta = dataB.mean - dataA.mean;
+  U count = dataA.count + dataB.count;
+  U mean, sigma2;
+  if (count > decltype(dataB.count){0}) {
+    auto coef = 1.f / count; // NB we don't use --use_fast_math, but this is
+                             // emulation, 1./count goes to intrinsic, `* coef`
+                             // is multiplication, instead of slow fp division
+    auto nA = dataA.count * coef;
+    auto nB = dataB.count * coef;
+    mean = nA * dataA.mean + nB * dataB.mean;
+    sigma2 = dataA.sigma2 + dataB.sigma2 + delta * delta * dataA.count * nB;
+  } else {
+    mean = U(0);
+    sigma2 = U(0);
+  }
+  return {mean, sigma2, count};
+}
+
+template <typename T, typename T_ACC, typename shared_t>
+WelfordDataLN compute_stats(
+    const T* __restrict__ X,
```

Comment:
```suggestion
    const T* RESTRICT X,
```
Otherwise it will fail on Windows.